src/ctx.cpp

```diff
@@ -1926,6 +1946,81 @@ llvm::Instruction *FunctionEmitContext::FPCastInst(llvm::Value *value, llvm::Typ
     return inst;
 }
 
+llvm::Value *FunctionEmitContext::__f2h_CastInst(const char *funcName, llvm::Value *v, llvm::Type *targetType,
+                                                 const llvm::Twine &name) {
+    if (v == NULL) {
+        AssertPos(currentPos, m->errorCount > 0);
+        return NULL;
+    }
+
+    llvm::Twine tw = name.isTriviallyEmpty() ? llvm::Twine(v->getName()) + "_f2h_cast" : name;
+
+    std::vector<Symbol *> foos;
+    m->symbolTable->LookupFunction(funcName, &foos);
+    assert(foos.size() == 2);
+    Symbol *s = foos[0];
+
+    const FunctionType *f0 = CastType<FunctionType>(foos[0]->type);
+    const FunctionType *f1 = CastType<FunctionType>(foos[1]->type);
+    if (targetType == LLVMTypes::Float16VectorType || targetType == LLVMTypes::FloatVectorType) {
+        if (f1->GetReturnType()->IsVaryingType()) {
+            s = foos[1];
+        } else if (f0->GetReturnType()->IsVaryingType()) {
+            s = foos[0];
+        } else {
+            assert(false);
+        }
+    } else {
+        if (f0->GetReturnType()->IsUniformType()) {
+            s = foos[0];
+        } else if (f1->GetReturnType()->IsUniformType()) {
+            s = foos[1];
+        } else {
+            assert(false);
+        }
+    }
+
```

Comment:
Calling stdlib function seems very fragile in terms of finding the right function (especially if they are overloaded by parameter). One way to fix it is to move ISPC implementation of `half_to_float` and `float_to_half` to builtins and call here this builtin function instead of stdlib. Need to discuss.

Reply:
May be we should, at the moment, I don't really understand the border between builtins and stdlib.

Reply:
Current implementation of finding the particular conversion function is safe. I don't see any benefits in having additional builtin written in ISPC other than allowing float16 operations when stdlib is not linked. Let's not introduce any additional entities where we keep parts of implementation of ISPC standard library - stdlib (written in ISPC) and builtins (written in LLVM IR) are more than sufficient. I'm OK to keep implementation in stdlib and have a warning if float16 operation is used and stdlib is not linked.